python/cudf/cudf/core/indexing_utils.py

```diff
@@ -0,0 +1,250 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, List, Tuple, Union
+
+from typing_extensions import TypeAlias
+
+import cudf
+from cudf.api.types import (
+    _is_scalar_or_zero_d_array,
+    is_bool_dtype,
+    is_integer,
+    is_integer_dtype,
+)
+from cudf.core import copy_types as ct
+
+
+# Poor man's algebraic data types
```

Comment:
Let's use another phrase or just drop the phrase to avoid being too colloquial. Also, what does it mean to have "algebraic data types"? I am familiar with the terms "sum types" or "product types" but I don't see the connection here.
```suggestion
# Algebraic data types
```

Reply:
Is the "algebraic" part that `IndexingSpec` is a sum type (variant of index-like types)?